# plm-utils
This repo contains a python package called `plmutils` that provides a set of tools for generating and analyzing embeddings of protein sequences using pre-trained protein language models (PLMs).

It was initially developed as a tool to identify sORFs for the [peptigate pipeline](https://github.com/Arcadia-Science/peptigate).

The code related to sORF prediction is confined to the `plmutils.tasks.classify_orfs` module. This includes the code to construct training and test datasets and the code to train and evaluate classification models for the task of predicting whether putative ORFs (of any length) are coding or noncoding.


## Installation
Create a new python 3.11 virtual env from the `dev.yml` file:
```bash
mamba env create -n plmutils-env -f dev.yml
```

Then activate the environment:
```bash
conda activate plmutils-env
```

Then install the `plmutils` package in editable mode:
```bash
pip install -e .
```

Finally, check that pytorch can find the GPU:
```bash
python -c "import torch; print(torch.cuda.is_available())"
```

## Usage
The main package defines several top-level generic commands:

- `plmutils translate`: This command uses `orfipy` to find putative ORFs in a fasta file of transcripts and translates them to protein sequences. The translated sequences are saved to a new fasta file. The `--longest-only` option can be used to only keep the longest ORF for each transcript.

- `plmutils embed`: This command uses a pre-trained protein language model to embed a fasta file of protein sequences. The resulting matrix of embeddings is saved as a numpy array in a `.npy` file. The order of the rows of this matrix corresponds to the order of the sequences in the input fasta file.

- `plmutils train`: This command trains a generic binary classifier given two embedding matrices, one for the positive class and one for the negative class. The trained classifier is optional saved to a user-specified directory.

- `plmutils predict`: This command generates predictions given an embedding matrix and a pre-trained classifier generated by the `train` command above.


## Coding/noncoding transcript prediction
The `plmutils.tasks.classify_orfs` module contains the code to train and evaluate a classifier to predict sORFs. This module defines its own CLI that is namespaced under the `plmutils classify-orfs` subcommand.
