# plm-utils
This repo contains a python package called `plmutils` that provides a set of tools for generating and analyzing embeddings of protein sequences using pre-trained protein language models (PLMs).

It was initially developed as a tool to identify sORFs for the [peptigate pipeline](https://github.com/Arcadia-Science/peptigate).

The code related to ORF prediction is confined to the `plmutils.tasks.orf_prediction` module. This includes the code to construct training and test datasets and the code to train and evaluate classification models for the task of predicting whether putative ORFs are coding or noncoding. Although this code was written for the specific task of predicting short ORFs (sORFs), it has no intrinsic dependence on ORF length. Indeed, the length of ORFs on which models are trained and evaluated is determined by a user-specified `--max-length` CLI option.


## Installation
Create a new python 3.11 virtual env from the `dev.yml` file:
```bash
mamba env create -n plmutils-env -f dev.yml
```

Then activate the environment:
```bash
conda activate plmutils-env
```

Then install the `plmutils` package in editable mode:
```bash
pip install -e .
```

Finally, check that pytorch can find the GPU:
```bash
python -c "import torch; print(torch.cuda.is_available())"
```

## Usage
The main package defines several generic (i.e., task-nonspecific) commands:

- `plmutils translate`: This command uses `orfipy` to find putative ORFs in a fasta file of transcripts and translates them to protein sequences. The translated sequences are saved to a new fasta file. The `--longest-only` option can be used to only keep the longest ORF for each transcript.

- `plmutils embed`: This command uses a pre-trained protein language model to embed a fasta file of protein sequences. The resulting matrix of embeddings is saved as a numpy array in a `.npy` file. The order of the rows of this matrix corresponds to the order of the sequences in the input fasta file.

- `plmutils train`: This command trains a generic binary classifier given two embedding matrices, one for the positive class and one for the negative class. The trained classifier is optional saved to a user-specified directory.

- `plmutils predict`: This command generates predictions given an embedding matrix and a pre-trained classifier generated by the `train` command above.


## Putative ORF prediction
The `plmutils.tasks.orf_prediction` module contains the code to train and evaluate classification models to predict ORFs from transcriptomic data. It implements one specific approach to this problem that uses extant annotated transcriptomes to obtain a set of putative ORFs that are likely to be "real" (i.e., from a coding transcript) and a set of ORFs that are likely to be "false" (i.e., from a noncoding transcript). These ORFs are then used to train a classifier to predict whether a given putative ORF is likely to be "real" or not.

The approach consists of the following steps:
1. Retrieve transcriptomes for a set of user-defined species from Ensembl.
1. Construct datasets of deduplicated coding and noncoding transcripts by clustering transcripts by sequence identity and retaining only one representative sequence from each cluster.
1. Identify and translate the longest putative ORF from each coding and noncoding transcript.
1. Generate the ESM embeddings of the translated ORFs.
1. Train a separate classification model on the embeddings from each species to predict whether a given ORF orginated from a coding or noncoding transcript.
1. Evaluate each trained model separately on each of the species to generate a matrix of model performance metrics for all pairs of species (one used to train the model, the other to evaluate it). This is intended to assess the extent to which the models generalize across species.

These steps are executed via a CLI defined in the `orf_prediction` module and namespaced under the `plmutils orf-prediction` subcommand. This CLI allows the user to optionally train and/or evaluate models only on putative short ORFs (sORFs) via a user-defined `--max-length` option.
